### DISCUSSION TITLE
Suricata Performance Issue

### DISCUSSION THREAD

**Discussion by wdhachfeld** [Security-Onion-Solutions/securityonion#6957] [Unsupported Versions] [answered]
Hi Everyone,

Looking for a bit of help on solving an issue with Suricata performance. I've been running SO (currently 2.3.91) on my home network since around May, 2021 without significant issues. But since the new year I've been seeing issue with Suricata having brief periods of spiking packet loss in the 30-40% range. The start of these packet loss spikes appears to have coincided with my implementation of a (Apple Time Machine) backup server for two of my laptops.

A quick description of my configuration...

Connection to the world at large is provided by a cable modem (~220 Mbps down, ~13 Mbps up) connected to a router/firewall (Ubiquiti Edge Router 4). The router's configured 1 Gbps LAN port is connected to a 10 Gbps switch (Ubiquiti Edge Switch 16 XG). The switch is configured to mirror all traffic on my primary VLAN to one of its ports. Connected to the switch is a custom built SuperMicro server using the Atom C3958 processor with 16 physical cores running at a flat (no turbo) 2 GHz. Server has 128 GB of RAM, 2 TB NVMe M.2 SSD, 2 TB SATA SSD, and 4 TB SATA SSD. The server has 2 x 10 Gbps SFP+ connections into the switch. One for general LAN traffic. And a second for the TAP port. The server is running VMware ESXi 7.0.2. And SO 2.3.91 is running in a VM on this ESXi server. Currently there is little else running on the sever. A UPS management VM and a 2nd low-CPu-use Linux VM.

Originally (prior to the performance issues) I had SO configured with 8 VCPU and 32 GB of RAM. The root drive is 256 GB on the NVMe M.2 SSD. And /nsm is the entire 4 TB SATA SSD. Suricata and Zeek were both configured with 2 threads each.

I've tried a number of different things based on the SO documentation to no avail.

Currently I have SO configured with 12 VCPU (pinned in ESXi to 12 physical CPUs) and 64 GB of RAM. Drives are the same as before. And I configured Suricata and Zeek as follows:
```
...
  suripins:
  - 4
  - 5
  - 6
  - 7
  - 8
  - 9
  - 10
  - 11
...
  zeek_pins:
  - 11
  - 10
  - 9
  - 8
  - 7
  - 6
  - 5
  - 4
...
```
Idea being to (I thought) give both Suricata and Zeek plenty of CPU to work with if they needed it. I verified that Suricata does, indeed, pick up the CPU configuration and use it. From `/var/logs/suricata/suricata.log`:
```
22/1/2022 -- 07:02:36 - <Notice> - This is Suricata version 6.0.4 RELEASE running in SYSTEM mode
22/1/2022 -- 07:02:36 - <Info> - CPUs/cores online: 12
22/1/2022 -- 07:02:36 - <Info> - dropped the caps for main thread
22/1/2022 -- 07:02:36 - <Info> - eve-log output device (regular) initialized: /nsm/eve-%Y-%m-%d-%H:%M.json
22/1/2022 -- 07:02:36 - <Info> - stats output device (regular) initialized: stats.log
22/1/2022 -- 07:02:36 - <Info> - Running in live mode, activating unix socket
22/1/2022 -- 07:02:39 - <Info> - 1 rule files processed. 19740 rules successfully loaded, 0 rules failed
22/1/2022 -- 07:02:39 - <Info> - Threshold config parsed: 13 rule(s) found
22/1/2022 -- 07:02:39 - <Info> - 19742 signatures processed. 1313 are IP-only rules, 3160 are inspecting packet payload, 15244 inspect application layer, 0 are decoder event only
22/1/2022 -- 07:03:13 - <Info> - Going to use 8 thread(s)
22/1/2022 -- 07:03:13 - <Info> - Running in live mode, activating unix socket
22/1/2022 -- 07:03:13 - <Info> - Using unix socket file '/var/run/suricata/suricata-command.socket'
22/1/2022 -- 07:03:13 - <Notice> - all 8 packet processing threads, 4 management threads initialized, engine started.
22/1/2022 -- 07:03:13 - <Info> - All AFP capture threads are running.
```
Here is what I see during one of the packet loss spikes:

![Suricata Packet Loss](https://user-images.githubusercontent.com/3442933/150655999-46775745-8309-498c-82e8-dba9210c37ee.png)
![Monitor Interface Traffic](https://user-images.githubusercontent.com/3442933/150656035-fc4637f7-e98b-4cc5-9262-c85a65672143.png)
![Overall CPU Usage](https://user-images.githubusercontent.com/3442933/150656301-c046266d-dfc3-4f23-991a-0c62d8cf8421.png)
![All Containers CPU Usage](https://user-images.githubusercontent.com/3442933/150656287-1a25d15f-bd59-4d86-bc66-13bdf617166f.png)
![Suricata Container CPU Usage](https://user-images.githubusercontent.com/3442933/150656064-db732653-6c0d-4158-ada3-35190fcd4160.png)

So the Suricata packet loss spike does appear to coincide with a jump in the monitor interface traffic. But it isn't **that** high at only 40-50 Mb/s. During this period there is zero Stenographer or Zeek packet loss. And Zeek capture loss in the 0.1-0.3% range. Note that the CPU usage of the `so-suricata` Docker container seems to clamp at 110%. A little more than 1 CPU. It is not using anywhere near the full 8 CPUs it has been given access to. In fact the system does not appear to be stressed for CPU at all. Disk I/O for the /nsm drive does not look especially high during this period either - in the neighborhood of 4 MiB.

Most of the rest of the time the Suricata packet loss is at zero. I.e. for the last 24 hours (including the narrow time window shown above):

![Suricata Packet Loss (24-Hour)](https://user-images.githubusercontent.com/3442933/150656555-341dc21f-4710-4c0f-873b-c4923fb2c5c1.png)

Anybody have any ideas? Thanks in advance for any suggestions!

**Comment by ccamejo1** [comment]
Have you tried pinning the CPU exclusively to either suri or zeek? From the config you posted it seems like you are pinning the same CPU to both.
Also, have you tried using the suriprocs, zeek_lbprocs instead of pinning and distribute the 8 amongst them? Then taking some from Zeek and adding to Suri to see if you notice improvements?  

C²

**Reply by wdhachfeld** [reply]
Hi C<sup>2</sup>,

Thanks for jumping in to help me out. Yes. At first when the original 2/2 (suriprocs/zeek_lbprocs) configuration "stopped" working I adjusted to 4/2 (no pinning). That didn't help. On a subsequent attempt I tried:
```
...
  suripins:
  - 4
  - 5
  - 6
  - 7
  - 8
  - 9
...
  zeek_pins:
  - 10
  - 11
...
```
Which also didn't help. The results were basically the same as the end configuration I showed above.

You are absolutely correct that in the posted configuration I am pinning both Suricata and Zeek to the same CPU set. With the later in reverse order. The idea being that maybe there was a prioritization scheme when it came to packet queueing/handling for either or both. I.e. allow each to use up to 8 CPUs as transient demand required. But (hopefully) only conflicting (doubling up) in the most extreme conditions.

Regardless of the configuration I've tried (pinned or non-pinned) I never seem to see the Suricata container CPU usage during one of these packet loss spikes to rise above the 120% range.

**Reply by ccamejo1** [reply]
wdhachfeld,

Anything unusual in /opt/so/log/suricata/suricata.log?  

Does flow.emerg_mode_entered in /opt/so/log/suricata/stats.log show anything > 0?

C²

**Reply by wdhachfeld** [reply]
Hey C<sub>2</sub>,

The Grafana graphs in my original post were from 2022-01-22. Nothing unusual in `/opt/so/log/suricata/suricata-20220122.log.gz`. The log contents in my original post are actually the complete contents of that log file.

I'm not familiar enough with the Suricatra architecture to know necessarily what qualifies as unusual in the reported statistics. The output of `gunzip -c /opt/so/log/suricata/stats-20220122.log.gz | grep "flow\.emerg_mode_entered"` does not show any counts greater than zero. Selecting a single set of counters that were reported near the end of one of the packet loss spikes shows the following non-zero values:
```
------------------------------------------------------------------------------------
Date: 1/22/2022 -- 00:02:36 (uptime: 0d, 05h 24m 37s)
------------------------------------------------------------------------------------
Counter                                       | TM Name                   | Value
------------------------------------------------------------------------------------
capture.kernel_packets                        | Total                     | 13114990
decoder.pkts                                  | Total                     | 13115351
decoder.bytes                                 | Total                     | 14579350164
decoder.ipv4                                  | Total                     | 12921204
decoder.ipv6                                  | Total                     | 164650
decoder.ethernet                              | Total                     | 13115351
decoder.tcp                                   | Total                     | 12227906
decoder.udp                                   | Total                     | 842075
decoder.icmpv4                                | Total                     | 2427
decoder.icmpv6                                | Total                     | 10376
decoder.avg_pkt_size                          | Total                     | 1111
decoder.max_pkt_size                          | Total                     | 1514
flow.tcp                                      | Total                     | 19521
flow.udp                                      | Total                     | 54544
flow.icmpv4                                   | Total                     | 238
flow.icmpv6                                   | Total                     | 1418
flow.tcp_reuse                                | Total                     | 2
flow.wrk.spare_sync_avg                       | Total                     | 100
flow.wrk.spare_sync                           | Total                     | 478
decoder.event.ipv4.opt_pad_required           | Total                     | 3070
decoder.event.ipv6.zero_len_padn              | Total                     | 2645
flow.wrk.flows_evicted_needs_work             | Total                     | 13092
flow.wrk.flows_evicted_pkt_inject             | Total                     | 15857
flow.wrk.flows_evicted                        | Total                     | 15272
flow.wrk.flows_injected                       | Total                     | 13052
tcp.sessions                                  | Total                     | 15921
tcp.pseudo                                    | Total                     | 220
tcp.syn                                       | Total                     | 16538
tcp.synack                                    | Total                     | 16562
tcp.rst                                       | Total                     | 18544
tcp.stream_depth_reached                      | Total                     | 457
tcp.overlap                                   | Total                     | 11167
detect.alert                                  | Total                     | 1
app_layer.flow.http                           | Total                     | 1088
app_layer.tx.http                             | Total                     | 1286
app_layer.flow.tls                            | Total                     | 14387
app_layer.flow.smb                            | Total                     | 1
app_layer.tx.smb                              | Total                     | 13
app_layer.flow.dns_tcp                        | Total                     | 5
app_layer.tx.dns_tcp                          | Total                     | 10
app_layer.flow.ntp                            | Total                     | 1721
app_layer.tx.ntp                              | Total                     | 2060
app_layer.flow.failed_tcp                     | Total                     | 15
app_layer.flow.dns_udp                        | Total                     | 34546
app_layer.tx.dns_udp                          | Total                     | 73388
app_layer.flow.failed_udp                     | Total                     | 18277
flow.mgr.full_hash_pass                       | Total                     | 81
flow.spare                                    | Total                     | 10042
flow.mgr.rows_maxlen                          | Total                     | 3
flow.mgr.flows_checked                        | Total                     | 111532
flow.mgr.flows_notimeout                      | Total                     | 52774
flow.mgr.flows_timeout                        | Total                     | 58758
flow.mgr.flows_evicted                        | Total                     | 58895
flow.mgr.flows_evicted_needs_work             | Total                     | 13052
tcp.memuse                                    | Total                     | 4849784
tcp.reassembly_memuse                         | Total                     | 6094848
http.memuse                                   | Total                     | 17041
flow.memuse                                   | Total                     | 8033984
```
Note the short up-time here is because I had reconfigured the thread counts again and restarted the Suricata and Zeek containers.

Do any of these counters suggest a specific issue?

One thing that occurs to me looking at this data... In particular the noted `app_layer.flow.smb` of 1. My speculation is that these packet losses are occurring during the backup of one or more of the 2 laptops to my Apple Time Machine backup server (a Mac Mini). I'm thinking this because the packet loss spikes began occurring right about the time I implemented the backups. That traffic would be SMB.

Do you know if Suricata requires all app-layer processing to go through a single packet processing thread? I.e. if all of the packets associated with a single backups session over SMB was required to go through a single Suricata packet processing thread it could explain the behavior I'm seeing. Most of the 40-50 Mb/s during the spike is probably associated with the backup. And if that traffic can't be split across multiple Suricata packet processing threads, it is possible that traffic is overwhelming a single thread running on a single 2 GHz core.

Thanks again, C<sub>2</sub>, for helping me figure this out!

**Reply by ccamejo1** [reply]
Hey wdhachfeld,

Glad to help.  Nothing in particular stands out to me in your stats, but others may have additional insight.  I don't have an answer for you on how Suricata handles app-layer processing.  I would like to think it wouldn't tie all app-layer processing to one thread, but I do not want to misinform you.

To your point about the Suricata CPU usage seeming low, I did come across this in the Suricata docs that reminded me of another setting you can try.  From my experience increasing the max-pending-packets field located in /opt/so/saltstack/defaults/salt/suricata/defaults.yaml (after copying to your /opt/so/saltstack/local/salt/suricata/) does directly increase CPU usage for Suricata and has decreased packet loss for us.

![image](https://user-images.githubusercontent.com/13799970/150699847-d3f1e22a-5819-4cd0-9a43-eeec8efb2dce.png)

SO2 defaults to 5000 out of the box.  You can try increasing that value in increments to see if you notice any improvement in your packet loss.  We currently have it set to 25000, but are using 42 suriprocs of a 48 core/128GB RAM sensor on a saturated 10Gb/s link.  Which is usually pegged at 9.76 Gb/s with mostly 0% Suricata packet loss.  I say mostly, because we do experience occasional sub 2% loss when traffic spikes above that 9.76 Gb/s threshold.
![image](https://user-images.githubusercontent.com/13799970/150700915-1e041bbf-4e42-4642-9e52-1cfaf8064764.png)

From the Suricata Docs:
"...Choosing a low number of packets being processed while having many CPU’s/CPU cores, can result in not making use of the whole computer-capacity. (For instance: using one core while having three waiting for processing packets.)"

https://suricata.readthedocs.io/en/suricata-6.0.0/configuration/suricata-yaml.html#max-pending-packets

This is one of many settings that can be adjusted to get Suri performance  improved.  SEPTun and SEPTun-Mark-II mentioned in the Security Onion docs https://docs.securityonion.net/en/2.3/performance.html?highlight=SEPTun have a wealth of information on Suricata performance tuning.
 
Hope this helps,

C²

**Reply by wdhachfeld** [reply]
Thanks for pointing out this `max-pending-packets` setting, @ccamejo1. This definitely appears to be the issue. Yesterday I doubled `max-pending-packets` from 5,000 to 10,000. Here is the result over the last 24 hours:

![Suricata Packet Loss](https://user-images.githubusercontent.com/3442933/151039125-20fc686c-ce91-4e7e-8adb-b3c5c5bfe13e.png)

There are still small spikes of packet loss. But they are now all under 10% where they had readily exceeded 30% before. I had no idea until your post that Suricata had a limit on the number of pending packets (although it does, of course, make much sense). Looking at the number of packets (rather than the amount of data) during the last 24 hours shows:

![Monitor Interface Packets](https://user-images.githubusercontent.com/3442933/151039458-42685c15-2486-4e1e-a9d3-54049f2ba823.png)

Definitely exceeding 5 (or even 10) kp/s at times. Interestingly even though the packet loss dropped significantly, the amount of CPU usage did not increase substantially:

![Suricata Container CPU Usage](https://user-images.githubusercontent.com/3442933/151039675-1e84699f-bbe7-4bda-8ffc-aed91cc991c0.png)

Still not going much above 120%.

I'll bump `max-pending-packets` up some more to see if that completely eliminates the packet loss spikes. And if it does, I'll have my answer. Will go back to my original configuration, and implement BPF disablement of Steno/Suricata/Zeek processing of the SMB traffic to the backup server.

**Comment by dougburks** [comment]
> The start of these packet loss spikes appears to have coincided with my implementation of a (Apple Time Machine) backup server for two of my laptops.

Have you considered creating a `nids` BPF so that Suricata will ignore traffic between the backup server and the two laptops?
https://docs.securityonion.net/en/2.3/bpf.html

**Reply by presianbg** [reply]
@wdhachfeld , do not forget to check those awesome videos on the topic - BPF filters:
https://www.youtube.com/watch?v=uamNOjtUi4Y
https://www.youtube.com/watch?v=H1azxPjDdYY

**Reply by wdhachfeld** [reply]
Thanks everyone.

I'm going to start with C<sub>2</sub>'s suggestion of increasing the `max-pending-packets` setting to see if that moves things. And I missed the mention of those SEPTun and SEPTun-Mark-II docs mentioned in the SO documentation. So I'll definitely look through those too.

If none of that helps, my next step will be to restore the original 2/2 suriprocs/zeek_lbprocs configuration. Go back to the original 8 VCPU and 32 GB VM configuration. And then do as Doug has suggested and disable the Suricata processing of the traffic to the backup server. This should help me confirm for certain that it is, in fact, the backups that are causing the packet loss spikes to occur.

And that might end up being the permanent solution to this. But I'd ideally like to achieve some level of understanding as to **why** I'm not able to process a meager 40-50 Mb/s with my configuration. If the CPU's were pegged, that would be one thing. Then those Atom processors I'm using (for lower TDP and fan noise) just aren't powerful enough. But that doesn't seem to be the case here.

I'll provide an update once I've been able to try a couple of these ideas. Thanks again everyone!

**Reply by presianbg** [reply]
Maybe the issue is not it the Mbps yard, but rather in the pps ? 
Even if you do not have issues (packet capture / processing loss), why would you need to have events / pcaps for a regular backup traffic ?

**Reply by wdhachfeld** [reply]
Hi @presianbg,

Fair enough. I don't know for certain if it is the quantity of data, or the number of packets, specifically that is causing the problem. I actually don't know for 100% certain that it is even the backups that are causing the problem. I'm speculating that based on the timing of when I implemented backups for the laptops and when the packet loss spikes began occurring.

And I would concur that I don't really need events / pcaps for regular backup traffic. In another discussion thread last week (#6911) Doug had suggested to me (as he has again above) the idea of simply using a BPF rule to ignore all traffic to the backup server. What I *would* ideally like is to monitor all *non-backup* traffic to/from the backup server. The reason I'm running SO on my home network is to keep an eye on all the devices that don't have active AV running on them. E.g. IoT devices. I would want to know, for example, if my thermostat was suddenly talking to the backup server. Because it certainly shouldn't be.

So as I noted earlier today, using a BPF rule to disable monitoring of the backup server may be the permanent solution. Maybe restricting the ignore rule to the SMB ports only for the backup server. That probably gives me about what I'm looking for.

But I'd like to understand the "why" if possible. Not just work around the issue. So that the next time I add another server or other devices to the network and something like this crops up I'll have a better idea what to do about it.

**Comment by wdhachfeld** [comment, marked as answer]
The cause of these Suricata packet loss spikes appears to be the increased influx of packets associated with the backup traffic (an "elephant flow" in Suricata documentation terms) that I added to the network. Combined with the relatively low default setting of Suricata's `max-pending-packets` configuration parameter. The Suricata documentation [suggests](https://suricata.readthedocs.io/en/suricata-6.0.0/performance/tuning-considerations.html) 10,000 or higher. But the default value for SO is currently 5,000.

I'm going to return my configuration to 8 VCPU, maybe leave the 64 GB memory, put back in 2/2 suriprocs/zeek_lbprocs, probably leave the `max-pending-packets` at 10,000. And then use BPF rules as suggested by @dougburks and @presianbg to suppress Stenographer, Suricata, and Zeek processing of the SMB traffic to the backup server. Which, incidentally, is also a recommendation in the Suricata documentation [here](https://suricata.readthedocs.io/en/suricata-6.0.0/performance/ignoring-traffic.html).

**Reply by dougburks** [reply]
I've updated https://docs.securityonion.net/en/2.3/suricata.html#performance with a little more detail.
